### PR TITLE
Precompile with SnoopCompile.jl

### PR DESCRIPTION
**Pull request by mtfishman**
This is an attempt to use precompilation to improve the time to compile functions. Basically, I just followed the instructions here: https://timholy.github.io/SnoopCompile.jl/stable/snoopi/

Here are some results. Without precompilation:
```julia
julia> @time using ITensors;
  1.386614 seconds (3.35 M allocations: 190.653 MiB, 0.97% gc time)

julia> @time i = Index(2);
  0.000674 seconds (23 allocations: 20.328 KiB)

julia> @time A = randomITensor(i,i');
  0.079560 seconds (222.12 k allocations: 11.910 MiB)

julia> @time A*A;
  1.051269 seconds (5.26 M allocations: 264.229 MiB, 6.51% gc time)

julia> @time svd(A,i);
  4.209746 seconds (18.30 M allocations: 908.040 MiB, 5.90% gc time)
```
and with precompilation:
```julia
julia> @time using ITensors;
  2.116096 seconds (4.31 M allocations: 271.163 MiB, 2.79% gc time)

julia> @time i = Index(2);
  0.000688 seconds (23 allocations: 20.328 KiB)

julia> @time A = randomITensor(i,i');
  0.074153 seconds (174.88 k allocations: 9.203 MiB)

julia> @time A*A;
  0.838833 seconds (2.83 M allocations: 153.644 MiB, 4.03% gc time)

julia> @time svd(A,i);
  1.708583 seconds (5.15 M allocations: 254.049 MiB, 3.27% gc time)
```
`svd` is improved a lot, which is nice. Note that the runtime of these functions are very small, so the times you are seeing are pretty much all compile times.

Basically how it works is that SnoopCompile.jl runs a specified file containing ITensor functions, creates a list of all function calls and the types of their inputs, and then generates a long list of calls to the `precompile(...)` function (found in `src/precompile/precompile.jl`). This is all done by the script `src/precompile/make_precompile.jl`. Julia uses that list of calls to `precompile` to precompile those functions.

### CODE REVIEW

**Comment by emstoudenmire** [comment]
Looks like a really big improvement! Great to have this

**Comment by mtfishman** [comment]
Unfortunately, compile times for `dmrg` are not improved much by this. Without precompiling we had:
```julia
julia> @time N = 10;
  0.000000 seconds

julia> @time sites = siteinds("S=1/2", N);
  0.060715 seconds (184.72 k allocations: 9.450 MiB)

julia> @time ampo = AutoMPO();
  0.000001 seconds (2 allocations: 96 bytes)

julia> @time for j=1:N-1
               ampo .+= ("Sz",j,"Sz",j+1)
               ampo .+= (0.5,"S+",j,"S-",j+1)
               ampo .+= (0.5,"S-",j,"S+",j+1)
             end;
  0.123681 seconds (160.85 k allocations: 9.050 MiB, 54.25% gc time)

julia> @time H = MPO(ampo,sites);
  3.385704 seconds (13.03 M allocations: 664.278 MiB, 5.34% gc time)

julia> @time psi0 = randomMPS(sites, 10);
  1.046497 seconds (3.87 M allocations: 195.673 MiB, 3.51% gc time)

julia> @time sweeps = Sweeps(2);
  0.000002 seconds (5 allocations: 432 bytes)

julia> @time maxdim!(sweeps, 10, 20);
  0.012399 seconds (23.40 k allocations: 1.202 MiB)

julia> @time cutoff!(sweeps, 1E-11);
  0.000001 seconds

julia> @time energy, psi = dmrg(H, psi0, sweeps; quiet=true);
 11.141052 seconds (37.77 M allocations: 1.859 GiB, 3.67% gc time)
```
while with precompiling we have:
```julia
julia> @time N = 10;
  0.000000 seconds

julia> @time sites = siteinds("S=1/2", N);
  0.090424 seconds (222.71 k allocations: 11.573 MiB, 10.00% gc time)

julia> @time ampo = AutoMPO();
  0.000002 seconds (2 allocations: 96 bytes)

julia> @time for j=1:N-1
               ampo .+= ("Sz",j,"Sz",j+1)
               ampo .+= (0.5,"S+",j,"S-",j+1)
               ampo .+= (0.5,"S-",j,"S+",j+1)
             end;
  0.047860 seconds (83.14 k allocations: 4.724 MiB)

julia> @time H = MPO(ampo,sites);
  2.016548 seconds (4.08 M allocations: 210.733 MiB, 1.92% gc time)

julia> @time psi0 = randomMPS(sites, 10);
  0.564354 seconds (404.31 k allocations: 21.562 MiB)

julia> @time sweeps = Sweeps(2);
  0.000002 seconds (5 allocations: 432 bytes)

julia> @time maxdim!(sweeps, 10, 20);
  0.011348 seconds (23.39 k allocations: 1.202 MiB)

julia> @time cutoff!(sweeps, 1E-11);
  0.000000 seconds

julia> @time energy, psi = dmrg(H, psi0, sweeps; quiet=true);
  9.862911 seconds (28.68 M allocations: 1.424 GiB, 2.82% gc time)
```
I'm not sure why the `dmrg` compile time isn't improved more (again note that the runtime is around 0.03 seconds, so what is shown is basically all compile time).

**Comment by mtfishman** [comment]
I think this is a good start, so I'll merge.
